contrib/pg_tde/documentation/docs/architecture.md

```diff
@@ -0,0 +1,403 @@
+The old key is kept as is at the same location, as it is potentially still needed to decrypt backups or other databases.
+
+It is also possible that the location of the keyring changes:
+either the service is moved to a new address, or the keyring has to be moved to a different keyring provider type. Both of these situations are supported by `pg_tde` using simple SQL functions.
+
+In certain cases the SQL API can't be used for these operations:
+if the server isn't running, and the old keyring is no longer available, startup can fail if it needs to access the encryption keys. 
+For these situations, `pg_tde` also provides command line tools to recover the database.
+
+
+### Internal key regeneration
+
+Internal keys for files (tables, indexes, etc.) are fixed once the file is created, there's no way currently to re-encrypt a file.
+
+There are workarounds for this, because operations that move the table data to a new file, such as `VACUUM FULL` or an `ALTER TABLE` that rewrites the file will create a new key for the new file, essentially rotating the internal key.
+This however means taking an exclusive lock on the table for the duration of the operation, which might not be desirable for huge tables.
+
+### Internal key storage
+
+Internal keys, and generally `pg_tde` metadata is kept in a single directory in `$PGDATA/pg_tde`. 
+In this directory each database has separate files, containing:
+
+* Encrypted internal keys and internal key mapping to tables
+* Information about the key providers
+
+There's also a special global section marked with the OID 607, which includes the global key providers / global internal keys.
+
+This is used by the WAL encryption, and can optionally be used by specific databases too, if global provider inheritance is enabled.
+
+### Key providers (principal key storage)
+
+Principal keys are stored on external providers.
+Currently pg_tde has 3 implementations:
+
+* A local file storage, intended for development and testing only
+* Hashicorp Vault
+* KMIP compatible servers
+
+In all cases, `pg_tde` requires a detailed configuration, including the address of the service and authentication information.
+With these details, all operations are done by `pg_tde` based on user operations:
+
+* When a new principal key is created, it will communicate with the service and upload a fresh key to it
+* When a principal key is required for decryption, it will try to get the key from the service
+
+### Sensitive key provider information
+
+Some of the keyring provider information (authentication details) is potentially sensitive, and it is not safe to store it together with the database in the `$PGDATA` directory, or even on the same server.
+
+For this purpose, `pg_tde` provides a mechanism where instead of directly specifying the parameter, users can specify an external service instead, from where it downloads the information.
```

Comment:
```suggestion
For this purpose, `pg_tde` provides a mechanism for users to specify an external service from where to download the information instead of directly setting the parameter.
```